File: internal/updater/updater.go
// Package updater implements keep-at's self-update command: check GitHub
// releases for tweedge/keep-at, download the asset matching the current
// OS/architecture, and replace the running binary.
package updater

import (
	"archive/tar"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
)

// ReleasesURL is GitHub's API endpoint for the latest stable (non-prerelease)
// keep-at release. Release assets are expected to be named
// keep-at_<GOOS>_<GOARCH>.tar.gz, each containing a single "keep-at"
// binary - see the build scripts. It's a var (not const) so tests can point
// it at an httptest server.
var ReleasesURL = "https://api.github.com/repos/tweedge/keep-at/releases/latest"

// ReleasesListURL lists all releases, newest first, including prereleases.
// Used when the user opts into beta versions: the first non-draft release
// in the list may be a prerelease, which is exactly what a beta update wants.
// A var for the same testability reason as ReleasesURL.
var ReleasesListURL = "https://api.github.com/repos/tweedge/keep-at/releases"

type release struct {
	TagName string  `json:"tag_name"`
	Draft   bool    `json:"draft"`
	Assets  []asset `json:"assets"`
}

type asset struct {
	Name               string `json:"name"`
	BrowserDownloadURL string `json:"browser_download_url"`
}

// LatestVersion returns the tag name of the latest GitHub release.
// includeBeta selects whether prerelease (beta) releases are eligible.
func LatestVersion(client *http.Client, userAgent string, includeBeta bool) (string, error) {
	rel, err := fetchLatestRelease(client, userAgent, includeBeta)
	if err != nil {
		return "", err
	}
	return rel.TagName, nil
}

// Apply downloads the release asset matching the current platform and
// replaces the binary at currentExecPath with it, preserving its file
// mode. includeBeta selects whether prerelease (beta) releases are
// eligible; the default (false) only considers stable releases. It
// downloads to a temp file first and only swaps the executable in
// afterward, so a failed or interrupted update leaves the existing binary
// untouched.
func Apply(client *http.Client, userAgent, currentExecPath string, includeBeta bool) (newVersion string, err error) {
	rel, err := fetchLatestRelease(client, userAgent, includeBeta)
	if err != nil {
		return "", err
	}

	assetName := fmt.Sprintf("keep-at_%s_%s.tar.gz", runtime.GOOS, runtime.GOARCH)
	var downloadURL string
	for _, a := range rel.Assets {
		if a.Name == assetName {
			downloadURL = a.BrowserDownloadURL
			break
		}
	}
	if downloadURL == "" {
		return "", fmt.Errorf("updater: no release asset named %s in %s", assetName, rel.TagName)
	}

	binary, err := downloadBinary(client, userAgent, downloadURL)
	if err != nil {
		return "", err
	}

	if err := replaceExecutable(currentExecPath, binary); err != nil {
		return "", err
	}

	return rel.TagName, nil
}

// fetchLatestRelease returns the release keep-at should update to. With
// includeBeta false it uses GitHub's /releases/latest endpoint, which
// already excludes prereleases. With includeBeta true it lists all releases
// and takes the newest non-draft one, which may be a prerelease.
func fetchLatestRelease(client *http.Client, userAgent string, includeBeta bool) (*release, error) {
	if !includeBeta {
		return fetchReleaseFromURL(client, userAgent, ReleasesURL)
	}

	req, err := http.NewRequest(http.MethodGet, ReleasesListURL, nil)
	if err != nil {
		return nil, fmt.Errorf("updater: building request: %w", err)
	}
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Accept", "application/vnd.github+json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("updater: fetching releases: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("updater: fetching releases returned %s", resp.Status)
	}

	var releases []release
	if err := json.NewDecoder(resp.Body).Decode(&releases); err != nil {
		return nil, fmt.Errorf("updater: parsing release metadata: %w", err)
	}
	for i := range releases {
		if releases[i].Draft {
			continue
		}
		return &releases[i], nil
	}
	return nil, fmt.Errorf("updater: no releases found")
}

// fetchReleaseFromURL fetches a single release object from releasesURL.
func fetchReleaseFromURL(client *http.Client, userAgent, releasesURL string) (*release, error) {
	req, err := http.NewRequest(http.MethodGet, releasesURL, nil)
	if err != nil {
		return nil, fmt.Errorf("updater: building request: %w", err)
	}
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Accept", "application/vnd.github+json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("updater: fetching latest release: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("updater: fetching latest release returned %s", resp.Status)
	}

	var rel release
	if err := json.NewDecoder(resp.Body).Decode(&rel); err != nil {
		return nil, fmt.Errorf("updater: parsing release metadata: %w", err)
	}
	return &rel, nil
}

// downloadBinary fetches a .tar.gz release asset and extracts the single
// "keep-at" binary from it into memory.
func downloadBinary(client *http.Client, userAgent, url string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("updater: building download request: %w", err)
	}
	req.Header.Set("User-Agent", userAgent)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("updater: downloading %s: %w", url, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("updater: downloading %s returned %s", url, resp.Status)
	}

	gz, err := gzip.NewReader(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("updater: opening gzip stream: %w", err)
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			return nil, fmt.Errorf("updater: %s did not contain a keep-at binary", url)
		}
		if err != nil {
			return nil, fmt.Errorf("updater: reading tar archive: %w", err)
		}
		base := filepath.Base(hdr.Name)
		if base != "keep-at" && base != "keep-at.exe" {
			continue
		}
		return io.ReadAll(tr)
	}
}

// replaceExecutable atomically swaps targetPath's contents for newBinary,
// preserving targetPath's existing file permissions.
func replaceExecutable(targetPath string, newBinary []byte) error {
	info, err := os.Stat(targetPath)
	mode := os.FileMode(0o755)
	if err == nil {
		mode = info.Mode()
	}

	dir := filepath.Dir(targetPath)
	tmp, err := os.CreateTemp(dir, ".keep-at-update-*")
	if err != nil {
		return fmt.Errorf("updater: creating temp file in %s: %w", dir, err)
	}
	tmpPath := tmp.Name()
	defer os.Remove(tmpPath) // no-op once the rename below succeeds

	if _, err := tmp.Write(newBinary); err != nil {
		tmp.Close()
		return fmt.Errorf("updater: writing new binary: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("updater: finalizing new binary: %w", err)
	}
	if err := os.Chmod(tmpPath, mode); err != nil {
		return fmt.Errorf("updater: setting permissions: %w", err)
	}
	if err := os.Rename(tmpPath, targetPath); err != nil {
		return fmt.Errorf("updater: replacing %s: %w", targetPath, err)
	}
	return nil
}

File: src/updater.rs
//! Self-update: check GitHub releases, download the matching asset,
//! atomically replace the running binary.
//! Linux-only asset naming: keep-at_linux_<arch>.tar.gz.

use anyhow::{Context, Result};

pub static RELEASES_URL: &str = "https://api.github.com/repos/tweedge/keep-at/releases/latest";
pub static RELEASES_LIST_URL: &str = "https://api.github.com/repos/tweedge/keep-at/releases";

#[derive(Debug, serde::Deserialize)]
struct Release {
    tag_name: String,
    #[serde(default)]
    draft: bool,
    #[serde(default)]
    assets: Vec<Asset>,
}

#[derive(Debug, serde::Deserialize)]
struct Asset {
    name: String,
    browser_download_url: String,
}

fn releases_url() -> String {
    std::env::var("KEEPAT_RELEASES_URL").unwrap_or_else(|_| RELEASES_URL.to_string())
}

fn releases_list_url() -> String {
    std::env::var("KEEPAT_RELEASES_LIST_URL").unwrap_or_else(|_| RELEASES_LIST_URL.to_string())
}

async fn fetch_release(client: &reqwest::Client, user_agent: &str, url: &str) -> Result<Release> {
    let resp = client
        .get(url)
        .header("User-Agent", user_agent)
        .header("Accept", "application/vnd.github+json")
        .send()
        .await
        .with_context(|| format!("fetching {url}"))?;
    let status = resp.status();
    if !status.is_success() {
        anyhow::bail!("fetching {url} returned {status}");
    }
    resp.json().await.context("parsing release metadata")
}

async fn fetch_latest(
    client: &reqwest::Client,
    user_agent: &str,
    include_beta: bool,
) -> Result<Release> {
    if !include_beta {
        return fetch_release(client, user_agent, &releases_url()).await;
    }
    let resp = client
        .get(releases_list_url())
        .header("User-Agent", user_agent)
        .header("Accept", "application/vnd.github+json")
        .send()
        .await
        .context("fetching releases")?;
    let status = resp.status();
    if !status.is_success() {
        anyhow::bail!("fetching releases returned {status}");
    }
    let releases: Vec<Release> = resp.json().await.context("parsing release metadata")?;
    releases
        .into_iter()
        .find(|r| !r.draft)
        .context("no releases found")
}

pub async fn latest_version(
    client: &reqwest::Client,
    user_agent: &str,
    include_beta: bool,
) -> Result<String> {
    Ok(fetch_latest(client, user_agent, include_beta)
        .await?
        .tag_name)
}

/// Asset arch in the release naming scheme (Go-style GOARCH, matching
/// scripts/build-release.sh): amd64/arm64/arm/386 — NOT Rust target_arch
/// (x86_64/aarch64), which names no published asset.
pub fn asset_arch() -> &'static str {
    match std::env::consts::ARCH {
        "x86_64" => "amd64",
        "aarch64" => "arm64",
        "arm" => "arm",
        "x86" => "386",
        other => other,
    }
}

fn asset_name() -> String {
    format!("keep-at_linux_{}.tar.gz", asset_arch())
}

/// Download the matching asset and replace current_exe atomically.
pub async fn apply(
    client: &reqwest::Client,
    user_agent: &str,
    current_exe: &std::path::Path,
    include_beta: bool,
) -> Result<String> {
    let rel = fetch_latest(client, user_agent, include_beta).await?;
    let want = asset_name();
    let url = rel
        .assets
        .iter()
        .find(|a| a.name == want)
        .map(|a| a.browser_download_url.clone())
        .with_context(|| format!("no release asset named {want} in {}", rel.tag_name))?;

    let resp = client
        .get(&url)
        .header("User-Agent", user_agent)
        .send()
        .await
        .with_context(|| format!("downloading {url}"))?;
    let status = resp.status();
    if !status.is_success() {
        anyhow::bail!("downloading {url} returned {status}");
    }
    let body = resp.bytes().await.context("reading download")?;

    let binary =
        extract_binary(&body).with_context(|| format!("extracting keep-at binary from {url}"))?;
    replace_executable(current_exe, &binary)?;
    Ok(rel.tag_name)
}

fn extract_binary(tar_gz: &[u8]) -> Result<Vec<u8>> {
    use flate2::read::GzDecoder;
    use std::io::Read;
    let gz = GzDecoder::new(tar_gz);
    let mut tar = tar::Archive::new(gz);
    for entry in tar.entries().context("reading tar archive")? {
        let mut entry = entry?;
        let path = entry.path().context("reading tar entry path")?.into_owned();
        let base = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
        if base == "keep-at" {
            let mut buf = Vec::new();
            entry.read_to_end(&mut buf)?;
            return Ok(buf);
        }
    }
    anyhow::bail!("archive did not contain a keep-at binary")
}

fn replace_executable(target: &std::path::Path, new_binary: &[u8]) -> Result<()> {
    use std::os::unix::fs::PermissionsExt;
    let mode = std::fs::metadata(target)
        .map(|m| m.permissions().mode())
        .unwrap_or(0o755);
    let dir = target.parent().context("executable has no parent dir")?;
    let tmp = dir.join(format!(".keep-at-update-{}", std::process::id()));
    std::fs::write(&tmp, new_binary).with_context(|| format!("writing {}", tmp.display()))?;
    std::fs::set_permissions(&tmp, std::fs::Permissions::from_mode(mode))?;
    std::fs::rename(&tmp, target).with_context(|| format!("replacing {}", target.display()))?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn asset_arch_matches_release_naming() {
        // Release assets use Go-style GOARCH (build-release.sh). A mismatch
        // here breaks self-update on every platform with zero test signal
        // otherwise (found live: x86_64 vs amd64 on the stable channel).
        assert_eq!(
            asset_arch(),
            match std::env::consts::ARCH {
                "x86_64" => "amd64",
                "aarch64" => "arm64",
                "arm" => "arm",
                "x86" => "386",
                other => other,
            }
        );
        assert!(asset_name().starts_with("keep-at_linux_"));
        assert!(asset_name().ends_with(".tar.gz"));
    }
}
